source/server/listener_manager_impl.cc

```diff
@@ -561,23 +562,58 @@ void ListenerManagerImpl::startWorkers(GuardDog& guard_dog) {
   }
 }
 
+void ListenerManagerImpl::stopListener(Network::ListenerConfig& listener,
+                                       std::function<void()> callback) {
+  const uint64_t listener_tag = listener.listenerTag();
+  const auto workers_pending_stop = std::make_shared<std::atomic<uint32_t>>(workers_.size());
+  for (const auto& worker : workers_) {
+    worker->stopListener(listener, [this, callback, workers_pending_stop, listener_tag]() {
+      if (--(*workers_pending_stop) == 0) {
+        server_.dispatcher().post([this, callback, listener_tag]() {
+          for (auto& listener : active_listeners_) {
+            if (listener->listenerTag() == listener_tag) {
+              listener->socket().close();
+            }
+          }
+          for (auto& listener : draining_listeners_) {
+            if (listener.listener_->listenerTag() == listener_tag) {
+              // Handle the edge case when new listener is added for the same address as the drained
+              // one. In this case the socket is shared between both listeners so one should avoid
+              // closing it.
+              bool is_used_by_active_listener = false;
+              for (auto& active_listener : active_listeners_) {
```

Comment:
I'm wondering if it would be better to track the current state on each listener in the add/update path vs. re-computing it here? WDYT? IMO it's probably OK to not handle extreme edge cases completely gracefully as long as it works (to some definition of working) and it's well commented?

Reply:
Would it be acceptable to have some additional map from socket to the state?

Reply:
Ultimately I would go for whatever has the least amount of extra/special case logic. I will defer to you on the best way of handling that. If you think what you have here is the best we can do that is fine also.